sw/host/tests/rom/e2e_bootstrap_entry/src/main.rs

```diff
@@ -0,0 +1,161 @@
+    )]
+    jedec_page: usize,
+
+    #[structopt(
+        long, parse(try_from_str=u8::from_str),
+        default_value = "0xEF",
+        help = "JEDEC ID of manufacturer",
+    )]
+    jedec_id: u8,
+
+    #[structopt(
+        long, parse(try_from_str=u8::from_str),
+        default_value = "0x08",
+        help = "JEDEC manufacturer product ID",
+    )]
+    jedec_product: u8,
+
+    #[structopt(
+        long, parse(try_from_str=u32::from_str),
+        default_value = "0x100000",
+        help = "Size of the internal flash",
+    )]
+    flash_size: u32,
+}
+
+fn test_bootstrap_entry(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    // TODO: Should really `opts.init.uart_params.create()` here, but we need to refactor
+    // BootstrapOptions first.
+    //let uart = opts.init.uart_params.create(&transport)?;
+    let uart = transport.uart("0")?;
+    // Look for the bootstrap signal from the ROM.
+    let mut console = UartConsole {
+        timeout: Some(opts.timeout),
+        exit_success: Some(Regex::new(r"bootstrap:1\r\n")?),
+        exit_failure: Some(Regex::new(r"BFV:")?),
+        ..Default::default()
+    };
+
+    transport.apply_pin_strapping("ROM_BOOTSTRAP")?;
+    let reset_delay = opts
+        .init
+        .bootstrap
+        .options
+        .reset_delay
+        .unwrap_or(Duration::from_millis(50));
+    transport.reset_target(reset_delay, true)?;
+
+    // Now watch the console for the exit conditions.
+    let mut stdout = std::io::stdout();
+    let result = console.interact(&*uart, None, Some(&mut stdout))?;
+    match result {
+        ExitStatus::ExitSuccess => {}
+        _ => {
+            bail!("FAIL: {:?}", result);
+        }
+    };
+    Ok(())
+}
+
+fn test_jedec_id(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let spi = transport.spi("0")?;
+    let id = SpiFlash::read_jedec_id(&*spi, 16)?;
+    log::info!("Read jedec id: {:x?}", id);
+    let mut index = 0;
+    while id[index] == 0x7f {
+        index += 1;
+    }
+    let page = index;
+    let manufacturer = id[index];
+    let product = id[index + 1];
+    let density = id[index + 2];
+
+    assert_eq!(page, opts.jedec_page);
+    assert_eq!(manufacturer, opts.jedec_id);
+    assert_eq!(product, opts.jedec_product);
+    assert_eq!(1u32 << density, opts.flash_size);
+    Ok(())
+}
+
+fn test_sfdp(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let spi = transport.spi("0")?;
+    let sfdp = SpiFlash::read_sfdp(&*spi)?;
+    log::info!("SFDP: {:#x?}", sfdp);
+
+    // We expect a SFDP header of version 1.10 with no additional phdrs
+    // except the mandatory JDEC phdr.
+    assert_eq!(sfdp.header.signature, 0x50444653);
+    assert_eq!(sfdp.header.minor, 0xa);
+    assert_eq!(sfdp.header.major, 0x1);
+    assert_eq!(sfdp.header.nph, 0x0);
+    assert_eq!(sfdp.header.reserved, 0xff);
+
+    // Check the parameters of the mandatory phdr.
+    assert_eq!(sfdp.phdr[0].id, 0x0);
```

Comment:
MSB of BFPT's ID should be `0xff`, right? See [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/spi_device.c;drc=d425d553f22422679d93f87551c3e49b6b920c1a;l=76) and [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/spi_device.c;drc=d425d553f22422679d93f87551c3e49b6b920c1a;l=419).

Reply:
This is a matter of spec version.  In older versions of JESD216, the ID is an 8-bit field at offset 0 of the phdr struct.

In newer versions, the ID is apparently a 16-bit field composed of the 8-bit field at offset 0 and the previously unused byte located at offset 7 of the phdr struct.

The parser in opentitantool is currently based on the JESD216 from 2011.